### PR TITLE
Slight refactoring of PFEGammaAlgo

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

While other PRs are working on PFAlgo, I still have some suggestions for PFEGammaAlgo. In this PR I propose to:

1. Have a `const` reference to the employed GBRForests as a member so one does not have to pass them as a function argument with the rather obscure name `hoc` (for "HeavyObjectCache"). Having `const` members is ok after all, I only have it with the mutable members.
2. Have a new output type for the PFAlgo, so one only has to use the algo by the `operator()` instead of retrieving all the results that are saved as mutable members of the algo via getter functions
3. Merge PFEGammaProducer.h with .cc

PFEGammaAlgo has now only half as many mutable members as before, so we are on a good track towards a nice thread safe algo there I think!

#### PR validation:

CMSSW compiles and local matrix tests pass.

#### if this PR is a backport please specify the original PR:

No backport intended.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc [hunk 1]
```diff
@@ -657,12 +657,20 @@ bool PFEGammaAlgo::isMuon(const reco::PFBlockElement& pfbe) {
...
+  // this starts off as an inclusive list of prototype objects built from 
+  // supercluster/ecal-driven seeds and tracker driven seeds in a block
+  // it is then refined through by various cleanings, determining the energy 
+  // flow.
+  // use list for constant-time removals
+  std::list<ProtoEGObject> refinableObjects;
+
   _splayedblock.clear();
-  _refinableObjects.clear();
+  refinableObjects.clear();
```
[the remark this answers is not in the document]
Comment (guitargeek) [reply RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: True! Thanks for reviewing

**Comment by guitargeek** [comment]
Hi, I followed your two suggestions and made a little additional change that makes sense to me: I took the primaryVertex out of the `cfg_` member, because I think the point of these "configuration structs" should be that they contain immutable configuration data only.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Code refactoring and simplification according to what described
- Jenkins tests pass and show no differences

**Comment by fabiocos** [comment]
+1

**Pull request merged**